UefiPayloadPkg/Library/PciSegmentInfoLibAcpiBoardInfo/PciSegmentInfoLibAcpiBoardInfo.c

```diff
@@ -51,22 +54,56 @@ RetrieveMultiSegmentInfoFromHob (
 
   *NumberOfRootBridges = PciRootBridgeInfo->Count;
 
+  PciSegments1 = NULL;
+
   Size         = PciRootBridgeInfo->Count * sizeof (PCI_SEGMENT_INFO);
-  mPciSegments = (PCI_SEGMENT_INFO *)AllocatePool (Size);
-  ASSERT (mPciSegments != NULL);
-  ZeroMem (mPciSegments, PciRootBridgeInfo->Count * sizeof (PCI_SEGMENT_INFO));
+  PciSegments1 = (PCI_SEGMENT_INFO *)AllocatePool (Size);
+  ASSERT (PciSegments1 != NULL);
+  ZeroMem (PciSegments1, PciRootBridgeInfo->Count * sizeof (PCI_SEGMENT_INFO));
 
   //
   // Create all root bridges with PciRootBridgeInfoHob
   //
   for (Index = 0; Index < PciRootBridgeInfo->Count; Index++) {
     if (UplSegmentInfo->SegmentInfo[Index].SegmentNumber == (UINT16)(PciRootBridgeInfo->RootBridge[Index].Segment)) {
-      mPciSegments[Index].BaseAddress = UplSegmentInfo->SegmentInfo[Index].BaseAddress;
+      PciSegments1[Index].BaseAddress = UplSegmentInfo->SegmentInfo[Index].BaseAddress;
+    }
+
+    PciSegments1[Index].SegmentNumber  = (UINT16)(PciRootBridgeInfo->RootBridge[Index].Segment);
+    PciSegments1[Index].StartBusNumber = (UINT8)PciRootBridgeInfo->RootBridge[Index].Bus.Base;
+    PciSegments1[Index].EndBusNumber   = (UINT8)PciRootBridgeInfo->RootBridge[Index].Bus.Limit;
+  }
+
+  // Above logic may cause different EndBusNumber from the exactly same Segment when multiple bridges sharing the same Segment.
+  // Merge the multiple bridge with the same Segment into the same brigde.
+
+  SegmentCount = 1;
+  Index2       = 0;
+  for (Index = 1; Index < PciRootBridgeInfo->Count; Index++) {
+    if (PciSegments1[Index].SegmentNumber != PciSegments1[Index - 1].SegmentNumber) {
+      SegmentCount++;
+      Index2 = Index;

```

Comment:
Since we cannot assume ordering of the same/different PciSegment root bridges, the sorting will be a little complicated.
Let's skip this fix from current PR and focusing on SMBIOS and ACPI in this PR. We can work on PciSegment sorting separately later.